skipruntime-ts/helpers/src/utils.ts

```diff
@@ -68,3 +72,118 @@ export function reactiveResponseHeader(
 ): [string, string] {
   return ["Skip-Reactive-Response-Token", JSON.stringify(reactiveResponse)];
 }
+
+/*****************************************************************************/
+// Joins
+/*****************************************************************************/
+
+type JoinObject = { value: Json; side: "left" | "right" };
+
+class AddJoinSideField extends OneToOneMapper<Json, Json, JoinObject> {
+  constructor(private joinSide: "left" | "right") {
+    super();
+  }
+
+  mapValue(v: Json, _key: Json): JoinObject {
+    if (typeof v !== "object") {
+      throw new Error(
+        "joinCollections only works on objects, not: " + JSON.stringify(v),
+      );
+    }
+    return { value: v, side: this.joinSide };
+  }
+}
+
+function haveDistinctFields(obj1: object, obj2: object): boolean {
+  const keys1 = new Set(Object.keys(obj1));
+  const keys2 = Object.keys(obj2);
+
+  for (const key of keys2) {
+    if (keys1.has(key)) {
+      return false;
+    }
+  }
+
+  return true;
+}
+
+function mergeObjects(object1: JoinObject, object2: JoinObject): Json {
+  const v1 = object1.value;
+  const v2 = object2.value;
+  if (typeof v1 !== "object") {
+    throw new Error(
+      "mergeObjects only works with objects, not: " + JSON.stringify(v1),
+    );
+  }
+  if (typeof v2 !== "object") {
+    throw new Error(
+      "mergeObjects only works with objects, not: " + JSON.stringify(v2),
+    );
+  }
+  if (!haveDistinctFields(v1, v2)) {
+    throw new Error(
+      "Objects don't have distinct fields: " +
+        JSON.stringify(v1) +
+        " " +
+        JSON.stringify(v2),
+    );
+  }
+  return { ...v1, ...v2 };
+}
+
+class MergeJoinFields implements Mapper<JoinObject, JoinObject, Json, Json> {
+  constructor(private allowMultipleValuesOnBothSides: boolean) {}
+
+  mapEntry(
+    key: Json,
+    values: NonEmptyIterator<JoinObject>,
+  ): Iterable<[Json, Json]> {
+    const result: [Json, Json][] = [];
+    let countLeft = 0;
+    let countRight = 0;
+    for (const value of values) {
+      if (value.side === "left") {
+        countLeft++;
+      }
+      if (value.side === "right") {
+        countRight++;
+      }
+    }
+    if (countLeft > 1 && countRight > 1) {
+      if (!this.allowMultipleValuesOnBothSides) {
+        throw new Error(
+          "More than one value detected on both sides for key: " +
+            JSON.stringify(key),
+        );
+      }
+    }
+    for (const value1 of values) {
+      if (value1.side === "left") {
+        for (const value2 of values) {
+          if (value2.side === "right") {
+            result.push([key, mergeObjects(value1, value2)]);
+          }
+        }
+      }
+    }
+    return result;
+  }
+}
+
+export function joinCollections<
+  K extends Json,
+  V1 extends Json,
+  V2 extends Json,
```

Comment:
Even if TS doesn't check it, I think it would be a documentation improvement to use JsonObject for the value types here.

Reply:
Agreed. Done.